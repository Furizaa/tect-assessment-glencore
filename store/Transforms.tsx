import * as React from 'react';
import { createAction } from './actionHelper';
import { TDictionaryId } from './Dictionaries';

export type TTransformId = string;

export interface ITransform {
  readonly id: TTransformId;
  readonly dictionaryId: TDictionaryId;
  readonly from: Array<string>;
  readonly to: string;
}

export interface ITransformState {
  list: Array<ITransform>;
  selectedId: TTransformId | null;
}

export type TTransformActions = Global.TActionsUnion<typeof transformActions>;

export type TTransformContext = React.Dispatch<TTransformActions>;

export interface ITransformRenderProp {
  readonly transforms: Array<ITransform>;
  readonly selectedTransform: ITransform | null;
  readonly dispatch: TTransformContext;
}

export interface ITransformProviderProps {
  children: Global.TRenderProp<ITransformRenderProp>;
  dictionaryId: TDictionaryId;
}

const TRANS_ADD = 'TRANS.ADD';
const TRANS_REMOVE = 'TRANS.REMOVE';
const TRANS_SELECT = 'TRANS.SELECT';

export const transformActions = {
  add: (item: ITransform) => createAction(TRANS_ADD, item),
  remove: (id: TTransformId) => createAction(TRANS_REMOVE, id),
  select: (id: TTransformId) => createAction(TRANS_SELECT, id),
};

export const TransformContext = React.createContext({});

export const transformReducer: React.Reducer<
  ITransformState,
  TTransformActions
> = (state, action) => {
  switch (action.type) {
    case TRANS_ADD:
      // Check if the `to` value already exists in one of the transforms of the
      // current dictionary id.
      const existingTo = state.list.find(
        transform =>
          transform.to === action.payload.to &&
          transform.dictionaryId === action.payload.dictionaryId
      );
      // If so (see above), then replace the transform with a copy of itself
      // which has the new `from` values concatinated to itself.
      if (existingTo) {
        return {
          ...state,
          list: [
            ...state.list.filter(transform => transform.id !== existingTo.id),
            {
              ...existingTo,
              from: [...existingTo.from, ...action.payload.from],
            },
          ],
        };
      }
      // `to` doesn't exists so add the new transfom to the list as-is.
      return { ...state, list: [...state.list, action.payload] };

    case TRANS_REMOVE:
      return {
        selectedId: null,
        list: state.list.filter(transform => transform.id !== action.payload),
      };

    case TRANS_SELECT:
      return {
        ...state,
        selectedId: action.payload,
      };

    default:
      return state;
  }
};

export const TransformProvider = ({
  children,
  dictionaryId,
}: ITransformProviderProps) => {
  const [state, dispatch] = React.useReducer(transformReducer, {
    selectedId: null,
    list: [],
  });
  const transforms = state.list.filter(
    transform => transform.dictionaryId === dictionaryId
  );

  // Find the selected transform by its id - only if the dictionary id also matches.
  const selectedTransform = state.selectedId
    ? state.list.find(
        transform =>
          transform.id === state.selectedId &&
          transform.dictionaryId === dictionaryId
      ) || null
    : null;

  return (
    <TransformContext.Provider value={dispatch}>
      {children({ transforms, selectedTransform, dispatch })}
    </TransformContext.Provider>
  );
};
