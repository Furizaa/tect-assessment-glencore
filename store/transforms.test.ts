import { transformReducer, transformActions } from './Transforms';

const initialState = {
  list: [],
  selectedId: null,
};

const mockTransform = {
  from: ['a'],
  to: '1',
  id: '1',
  dictionaryId: '1',
};

describe('Transform store', () => {
  describe('reducer', () => {
    it('adds transformation', () => {
      const state = transformReducer(
        initialState,
        transformActions.add(mockTransform)
      );
      expect(state.list.length).toBe(1);
      expect(state.list[0]).toMatchSnapshot();
    });

    it('doesnt select a default transform', () => {
      const state = transformReducer(
        initialState,
        transformActions.add(mockTransform)
      );
      expect(state.selectedId).toBeNull();
    });

    it('selects a transform', () => {
      const state = transformReducer(
        {
          list: [mockTransform],
          selectedId: null,
        },
        transformActions.select('1')
      );
      expect(state.selectedId).toBe('1');
    });

    it('merges transforms with the same TO value and dictionary', () => {
      const state = transformReducer(
        {
          list: [mockTransform],
          selectedId: null,
        },
        transformActions.add({
          from: ['b'],
          to: '1',
          id: '2',
          dictionaryId: '1',
        })
      );
      expect(state.list[0].from).toEqual(['a', 'b']);
    });

    it('doesnt merges transforms with the same TO value and different dictionaries', () => {
      const state = transformReducer(
        {
          list: [mockTransform],
          selectedId: null,
        },
        transformActions.add({
          from: ['b'],
          to: '1',
          id: '2',
          dictionaryId: '2',
        })
      );
      expect(state.list[0].from).toEqual(['a']);
      expect(state.list[1].from).toEqual(['b']);
    });

    it('removes transforms', () => {
      const state = transformReducer(
        {
          list: [mockTransform],
          selectedId: null,
        },
        transformActions.remove('1')
      );
      expect(state.list.length).toBe(0);
    });
  });
});
