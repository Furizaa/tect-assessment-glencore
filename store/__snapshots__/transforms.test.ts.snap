// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transform store reducer adds transformation 1`] = `
Object {
  "dictionaryId": "1",
  "from": Array [
    "a",
  ],
  "id": "1",
  "to": "1",
}
`;
